src/tactic/vampire/model.lean

```diff
@@ -0,0 +1,67 @@
+/-
+  Copyright (c) 2019 Seul Baek. All rights reserved.
+  Released under Apache 2.0 license as described in the file LICENSE.
+  Author: Seul Baek
+
+  Models and variable assignments.
+-/
+
+import logic.function tactic.basic tactic.interactive
+import tactic.vampire.misc
+
+universe u
+
+variable {α : Type u}
+
+namespace vampire
+
+local notation f `₀↦` a := assign a f
+
+def value (α : Type u) : Type u := list α → (α × Prop)
+
+def value.default [inhabited α] : value α :=
+λ as, (default α, false)
+
+def evaluate (a : α) : value α :=
+λ _, (a, false)
+
+def denote (v : value α) : α := (v []).fst
+
+postfix  `ₑ` : 1000 := evaluate
+postfix  `ᵈ` : 1000 := denote
+
+lemma denote_evaluate (a : α) : (a ₑ)ᵈ = a := rfl
+
+@[reducible] def value.app (v w : value α) : value α :=
+λ as, v (wᵈ :: as)
+
+local infix `⬝` := value.app
+
+def model (α : Type u) : Type u := nat → value α
+
+def model.default (α : Type u) [inhabited α] : model α := λ _, value.default
+
+@[reducible] def vas (α : Type u) : Type u := nat → α
+
+def vas.default (α : Type u) [inhabited α] : vas α := λ _, default α
+
+lemma assign_app_evaluate_denote (M : model α) (v w : value α) :
+  (M ₀↦ v ⬝ wᵈₑ) = (M ₀↦ v ⬝ w) :=
+by ext k as; cases k; refl
```

Comment:
I think you can just `ext k as; refl` here?

Reply:
In fact, `v ⬝ wᵈₑ = v ⬝ w` is a theorem and `assign_app_evaluate_denote` becomes trivial.

Reply:
You're right. I deleted the lemma.